src/plugins/intel_cpu/src/graph.cpp

```diff
@@ -1464,66 +1466,60 @@ void Graph::Infer(SyncInferRequest* request) {
 void Graph::SortTopologically() {
     OV_ITT_SCOPE(FIRST_INFERENCE, itt::domains::intel_cpu_LT, "Graph::SortTopologically");
 
-    auto sort = [](const std::vector<NodePtr>& nodes) {
-        std::unordered_set<NodePtr> visited;
-        visited.reserve(nodes.size());
+    // Set execIndex of all nodes to default invaild value
+    for (auto &node : graphNodes) {
+        node->execIndex = -1;
+    }
+
+    auto sort = [this](const std::vector<NodePtr>& nodes) {
         std::vector<NodePtr> sorted;
         sorted.reserve(nodes.size());
 
+        int execIndexCnt = -1;
+
         std::function<void(const NodePtr)> visit;
-        visit = [&visited, &sorted, &visit](const NodePtr node) {
-            const bool inserted = visited.insert(node).second;
-            if (!inserted)
+        visit = [&execIndexCnt, &sorted, &visit](const NodePtr node) {
+            if (node->execIndex >= 0)
                 return; // already visited
 
             if (!node->parallelWith.empty()) {
                 for (auto& n : node->parallelWith) {
-                    for (size_t i = 0; i < n->getChildEdges().size(); i++) {
-                        visit(n->getChildEdgeAt(i)->getChild());
+                    for (size_t i = 0; i < n->getParentEdges().size(); i++) {
+                        visit(n->getParentEdgeAt(i)->getParent());
                     }
                 }
 
-                // make sure parallel nodes are always enqueue together
+                // parallel nodes has same execIndex
+                // so they can provide correct start/end time point for
+                // their input and output memory object
+                ++execIndexCnt;
                 for (auto& n : node->parallelWith) {
-                    if (std::find(sorted.begin(), sorted.end(), n) == sorted.end()) {
-                        sorted.push_back(n);
-                    }
+                    sorted.push_back(n);
+                    n->execIndex = execIndexCnt;
                 }
             } else {
-                for (size_t i = 0; i < node->getChildEdges().size(); i++) {
-                    visit(node->getChildEdgeAt(i)->getChild());
+                for (size_t i = 0; i < node->getParentEdges().size(); i++) {
+                    visit(node->getParentEdgeAt(i)->getParent());
                 }
 
                 sorted.push_back(node);
+                node->execIndex = ++execIndexCnt;
             }
         };
 
+        // Always start from output nodes
+        for (auto&& kvp : outputNodesMap) {
+            visit(kvp.second);
+        }
+
```

Comment:
@maxnick FYI.
Ideally, we should start from all the end-point nodes we have, including memory states (as it is done in ov core IR). But I guess after recent changes it is not really possible, since memory states container is global and include inner graphs